### PR TITLE
Add a XAML Theming spec

### PR DESCRIPTION
**Pull request by zadjii-msft**
## Summary of the Pull Request

This PR aims to outline my vision for what "XAML theming" might mean for the Windows Terminal in v2.0 and beyond. Now that #3789 has merged into master, I figured it was time to discuss what customizing other elements of the UI might look like as well. I've tried to include as many configurable elements of the UI as I could, referencing a ton of threads across the repo. 

## References
* #3327 - the original theming thread
* #3789 - Added support for the tab color picker

## PR Checklist
* [x] Specs #3327
* [x] I work here
* [x] Is documentation

## Detailed Description of the Pull Request / Additional comments

_read the spec_

## TODO

* [ ] What style syntax should we have for theme members? `element.property` or `elementProperty`?
* [ ] This is the worst "Future Considerations" section I've ever written, so it might need some elaborating upon.
* [ ] For future considerations: Should we allow the "set tab color" action to set the tab color to the terminal background? What value would that be?
* [ ] How do we describe acrylic brushes? (I want the titlebar to be acrylic, 50% opacity, hue `#ff00ff`)? What about other potential materials? (vintage opacity?)
* [ ] Does the user have to manually specify the color for the titlebar, and the find dialog BG, and the command palette BG, and the .... or is there some way they can just set all at once?

### CODE REVIEW

**Comment by mdtauk** [comment]
Can there be support for the Accent Color brushes, so if I were to set my cursor or selction to my Accent Color, as it changes, that change will be reflected in Windows Terminal?

**Comment by zadjii-msft** [comment]
> 
> 
> Can there be support for the Accent Color brushes, so if I were to set my cursor or selction to my Accent Color, as it changes, that change will be reflected in Windows Terminal?

So that might need to be it's own separate feature request. The cursor and the selection are drawn in DirectX down in `TermControl`, so they work differently than a lot of the things in this spec (which are all XAML elements). 

For those, we'd need to accept something like `"cursorColor": "accent"`. When we did that, `CascadiaSettings` would know that the string isn't a "#rrggbb" color, so it could look up the color when we're building the `TerminalSettings` to pass to the `TermControl`. However, after that point, if the user changes their accent color, that wouldn't be reflected dynamically in the Terminal, until the user manually refreshes their settings.

It's not a horrible idea, but probably something we should move to it's own thread for discussion, since I don't think it really fits in with a lot of the other things in this spec.

**Comment by mdtauk** [comment]
@zadjii-msft  I just assumed the ability to specify Xaml resources, would include the app getting the Accent Color resource to use for the UI.  I do see that with the console rendering being a separate system, passing any Xaml brushes through, would be a different issue.

File: doc/specs/#3327 - Application Theming/#3327 - Application Theming.md [hunk 1]
```diff
@@ -0,0 +1,500 @@
...
+            "applicationTheme": "light",
+            "colorScheme": "Solarized Light",
+            "tabIcon": "hidden",
+            "tabCloseButton": "hover"
+        }
+    ]
+}
+```
+
+> **TODO FOR DISCUSSION**: I've given both a `tab.<property>` and a
```
Comment (carlos-zamora) [inline doc/specs/#3327 - Application Theming/#3327 - Application Theming.md:133]: I personally like the `<element>.<property>` representation more. I think it's more clear.

I'm probably in the minority here but I also _love_ nested objects too. So, is there any consideration for...
```
<element>: 
{
  <property1>: <value1>,
  <property2>: <value2>
}
```

File: doc/specs/#3327 - Application Theming/#3327 - Application Theming.md [hunk 2]
```diff
@@ -0,0 +1,500 @@
...
+  terminal instance.
+* `terminalForeground` to use the default foreground color of the active
+  terminal instance.
+* `key:SomeXamlKey` to try and look `SomeXamlKey` up from our resources as a
+  `Color`, and use that color for the value.
+    - TODO DISCUSSION: Does anyone want this?
+    - is `accent` just `key:SystemAccentColor`? If it is, is it a reasonable
+      alias that we'd want to provide anyways?
+    - TODO DISCUSSION: PR[#5280] suggested `{ "key": "SomeResourceKey" }` for
+      string resources, should we use that format for colors like this as well?
```
Comment (carlos-zamora) [inline doc/specs/#3327 - Application Theming/#3327 - Application Theming.md:202]: Why `key:SomeXamlKey` as opposed to `SomeXamlKey` directly? We can just add valid enums. Then extensions can add more enums too. Maybe I'm just not understanding this part.

File: doc/specs/#3327 - Application Theming/#3327 - Application Theming.md [hunk 3]
```diff
@@ -0,0 +1,500 @@
...
+```json
+    "tab.background": {"single": "terminalBackground", "multiple": null},
+    "tab.background": {"single": "terminalBackground", "multiple": "#ff0000"},
+    "tab.background": [ "terminalBackground", null ]
+```
+
+ Also shown is an array based implementation, as an option. Overall I'm not
+ happy with this, so I think it shouldn't be in the final draft of the spec, but
+ I'm leaving it for now as an option.
+-->
```
Comment (carlos-zamora) [inline doc/specs/#3327 - Application Theming/#3327 - Application Theming.md:229]: _This alone_ could get its own spec hahaha. I think following the active terminal instance is good for now. I'm thinking of some weird behaviors we could do that I think just isn't worth dealing with in v1:
| Scenario | Behavior |
|--|--|
| vertical stack of panes | (1) inherit top color |
| horizontal stack of panes | (1) blend colors touching titlebar, (2) blend at pane separator |

File: doc/specs/#3327 - Application Theming/#3327 - Application Theming.md [hunk 4]
```diff
@@ -0,0 +1,500 @@
...
+  `terminalBackground` that lets users say "I want to use the active tab color
+  as the titlebar color".
+  - During [#3789], there was a point where the terminal raised actually
+    implemented something like this. In it's implementation, the titlebar color
+    would be slightly lighter or darker than the tab color (to provide some
+    contrast). We'd want to make sure that the user could specify both "I want
+    to use the tab color with some contrast applied" or just literally "Use
+    whatever the active tab's color is."
+
+### Default Themes
```
Comment (carlos-zamora) [inline doc/specs/#3327 - Application Theming/#3327 - Application Theming.md:348]: (Maybe worth mentioning in future considerations over v1) Should we consider a "defaults/list" structure like we did with `profiles.defaults`? For now, I'm guessing themes won't be too big, but over time that could become an issue (maybe)

File: doc/specs/#3327 - Application Theming/#3327 - Application Theming.md [hunk 5]
```diff
@@ -0,0 +1,500 @@
...
+theme that's not particularly appealing, they can always change it back.
+
+### Branding
+
+Are we concerned that by enabling theming, the appearance of the Terminal won't
+be as static, and won't necessarily have as specific a look? It might be harder
+for potential users see a screenshot of the Terminal and _know_ "Thats the
+Windows Terminal". Is this something we're really all that concerned about
+though? If this is something users want (it is), then shouldn't that be what
+matters?
```
Comment (carlos-zamora) [inline doc/specs/#3327 - Application Theming/#3327 - Application Theming.md:466]: I think that's where the defaults matter most (I'm guessing that's a conversation we'll have at some point too)

File: doc/specs/#3327 - Application Theming/#3327 - Application Theming.md [hunk 6]
```diff
@@ -0,0 +1,500 @@
...
+
+[TODO]: # TODO: We should include more mockups here. That would be nice.
+
+![Tab matches Terminal background](Tab-Matches-Terminal-Color-000.png)
+_fig 1_: Using a tab color set to "terminalBackground". The Windows PowerShell
+tab has also set its color with the color picker.
+
+![Acrylic in the titlebar](AcrylicInTitlebar.png)
+_fig 2_: Using an acrylic titlebar color, with a tab color set to
+"terminalBackground"
```
Comment (carlos-zamora) [inline doc/specs/#3327 - Application Theming/#3327 - Application Theming.md:409]: This gives me chills of the video from Build 2019 haha. I love it.

File: doc/specs/#3327 - Application Theming/#3327 - Application Theming.md [hunk 7]
```diff
@@ -0,0 +1,500 @@
...
+* `pane.borderColor`: Control the color of the border used to separate panes.
+  This is the color of the inactive border between panes.
+* `pane.activeBorderColor`: Control the color of the border of the active pane
+* `pane.borderWidth`: Control the width of the borders used to separate panes.
+* `window.applicationTheme`: If set, will set the XAML `RequestedTheme`
+  property. This can be one of `light`, `dark` or `system`. This controls how
+  XAML fundamentally styles UI elements. If not provided, will use the default
+  value "system", which will use whatever the system's default theme is.
+
+#### Theme Colors
```
Comment (zadjii-msft) [inline doc/specs/#3327 - Application Theming/#3327 - Application Theming.md:186]: Hey something else we should consider - someone might want an acrylic titlebar, that's _not_ the terminal BG color. This spec doesn't cover that case quite yet.

maybe `"property": { "color": "#rrggbb", "opacity": .5 }`?

File: doc/specs/#3327 - Application Theming/#3327 - Application Theming.md [hunk 8]
```diff
@@ -0,0 +1,500 @@
...
+* [ ] Feature Request: Setting to hide/remove close ("x") button from tabs ([#3335])
+* [ ] Various different tab sizing modes
+    - the current sizing, which is `SizeToContent`
+    - Setting a min/max width on tabs
+    - Configuring tabs to split the available space
+
+Other lower-priority ideas:
+* [ ] Enable hiding the tab icon altogether
+* [ ] Enable forcing tab icons to monochrome
+* [ ] Tab row height
```
Comment (michael-hawker) [inline doc/specs/#3327 - Application Theming/#3327 - Application Theming.md:48]: 🦙 The Tab width behavior would fit in here as well (Compact, Equal, SizeToContent).

Comment (zadjii-msft) [reply doc/specs/#3327 - Application Theming/#3327 - Application Theming.md:48]: Oh, that's a good point. We'll need to come up with a way of handling what happens when the user has it set in both the `theme` and directly in the global settings - probably makes sense to just use the global value to override the theme value, right?

Comment (michael-hawker) [reply doc/specs/#3327 - Application Theming/#3327 - Application Theming.md:48]: @zadjii-msft or does it make sense to just move the setting itself into the theme (and have some default value if the theme doesn't provide one)? Or use the 'global' setting as the fallback value if the theme doesn't contain one?

File: doc/specs/#3327 - Application Theming/#3327 - Application Theming.md (hunk 1 above)
Comment (zadjii-msft) [reply doc/specs/#3327 - Application Theming/#3327 - Application Theming.md:133]: ew, I hate the nested objects so much that I didn't even put them in here 😝 My preference is definitely leaning towards `element.property`

File: doc/specs/#3327 - Application Theming/#3327 - Application Theming.md (hunk 2 above)
Comment (zadjii-msft) [reply doc/specs/#3327 - Application Theming/#3327 - Application Theming.md:202]: I'm missing the thread it was mentioned in now, but someone once suggested using `"xaml:SomeXamlKey"`, and I really like that. 

I think the "object" notation for looking up a key like this made it easier for us to say "this is an object, it's definitely not a string value". You're right though, these are just enums, so we can manually make sure that `xaml:` isn't used as a prefix for any of them.

File: doc/specs/#3327 - Application Theming/#3327 - Application Theming.md (hunk 4 above)
Comment (zadjii-msft) [reply doc/specs/#3327 - Application Theming/#3327 - Application Theming.md:348]: See, I don't think we need a "defaults" for themes. People will only ever have one theme active at a time, for the entire window, so we don't really need to abstract that away. 

The only reason I've got a whole list of themes is so: 
* In the future, people can use extensions to install multiple themes, and just pick one they like
* People can iterate on ideas they like, and keep around old "themes" they create, even if they're using something else currently. 

I think it's fine for each theme to be it's own atomic object, not composed of a (user theme defaults)+(changes to the user theme defaults)

**Comment by mdtauk** [comment]
I posted this in the UI Mega Thread, but was trying to trackdown the current issue handling Acrylic and Titlebars, so I am posting this here also...

Ignoring the Rounded Window Border (that will probably come to Windows with the new "Sun Valley" UI refresh) Here is how I would imagine a more immersive Terminal could look.
image
image

So the Titlebar uses the Console background - the Tabs blend in with an active tab standing out - and the Window Controls adapting their colour to contrast well.

**Comment by Shomnipotence** [comment]
As discussed in the following issue, there are too many ideas to achieve, including but not limited to the title bar, the current label, the hover label, the background color, and the different settings for the title bar. There is no doubt that it will eventually Import to build a theme system. My confusion lies in whether such a "grand" personalized setting system is needed for WindowsTerminal? I think I can sort out which parts can be prioritized (although Microsoft may not be interested in this).
＃1963，＃1337，＃3061 / ＃3062，＃2994

The Fluent Design System is constantly evolving, it is always in development, and less than 1/3 of the system applications adopt them, so Windows will never have a unified user interface, because the speed of applications using the latest design language will never keep up The speed of design updates. Moreover, before the release of WinUI3, I am afraid that all Win32 applications will not have much design changes-this is because Microsoft seems to be hesitant on whether to use XAML to build new applications.

Based on this, we discuss the design of Windows Terminal to minimize design changes and be compatible with existing designs as much as possible.

Windows Terminal allows you to set the background, but when the background uses acrylic, you will encounter an extremely ugly title bar (tab page bar), so the first consideration is to extend the acrylic effect to the title bar (tab page bar).

image

Here is a question, how to deal with the current label color, I personally think that the label color and the background of the current page can be consistent. If the current background color is a picture, the theme color will be automatically extracted from the picture.
If you do this, you only need to specify two colors:

Title bar color
Window background color (=label color)
WindowsTerminal1
WindowsTerminal2
WindowsTerminal5


There is another solution here, which I saw in the Files App, using a completely floating tab design.

In this case, the color of the title bar (tab page bar) is the same as the background color of the window. If a picture is used, the picture is tiled or full in the current window. Add a rounded rectangle background to the current label to show the selection, as shown in the figure.

WindowsTerminal3
WindowsTerminal4
WindowsTerminal6


I think these adjustments will not increase too much workload and are in harmony with Fluent Design. I want Microsoft to work as quickly as possible. I really look forward to Sun Valley, but so far, I think the biggest changes may be the "alarm and clock app" and the news flow of the taskbar.

**Comment by Stanzilla** [comment]
@Shomnipotence love your first solution, that is pretty much perfect.

**Comment by gabrielconl** [comment]
There should be a tab.backgroundMode that sets when should the color be visible (all the time, hover+selection, selection).

**Comment by orcmid** [comment]
When there are two tabs, my eye goes to the prominent unselected tab.  Determination of the foreground tab is too subtle in that case.  If there were 3 tabs it might be more obvious about what tab is in the foreground.  I suspect my eye can be trained, but that is worrisome.  

Is there a way out of this that isn't complexifying?

**Comment by zadjii-msft** [comment]
I'm closing this for now. There's been no movement for a while, and we're not planning on landing this inside 2.0 anymore. Plus, #10509 definitely complicates the design a bit. Might as well not leave this stale PR open for so long. I'll come back to it in the future.

**Comment by PuruVJ** [comment]
Hi is this implemented yet? 😅

**Comment by zadjii-msft** [comment]
@PuruVJ Nope. This was a proposal for how we might want to do this in the future, but I've temporarily closed this out because
1. the team won't have enough time to give it a proper review anytime soon
2. it needs a little reworking for Mica (#10509).

This is my pet project for 3.0, but alas, it didn't end up making the cut for 2.0.

**Comment by PuruVJ** [comment]
Thanks!! Makes sense

**Pull request closed without merging**